### PR TITLE
REGR: Arrow backed objects not propagating exceptions

### LINKED ISSUE #54944: BUG: Overflow in arrow comparison operator hides actual error behind NotImplementedError
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
>>> a = pd.Series([1 << 63], dtype="uint64[pyarrow]")
>>> b = pd.Series([None], dtype="int64[pyarrow]")
>>> a < b
Traceback (most recent call last):
  File "C:\ps_311_arrow_13_env\Lib\site-packages\pandas\core\arrays\arrow\array.py", line 630, in _cmp_method
    result = pc_func(self._pa_array, self._box_pa(other))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ps_311_arrow_13_env\Lib\site-packages\pyarrow\compute.py", line 245, in wrapper
    return func.call(args, None, memory_pool)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pyarrow\_compute.pyx", line 367, in pyarrow._compute.Function.call
  File "pyarrow\error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow\error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Integer value 9223372036854775808 not in range: 0 to 9223372036854775807

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\ps_311_arrow_13_env\Lib\site-packages\pandas\core\ops\common.py", line 76, in new_method
    return method(self, other)
           ^^^^^^^^^^^^^^^^^^^
  File "C:\ps_311_arrow_13_env\Lib\site-packages\pandas\core\arraylike.py", line 48, in __lt__
    return self._cmp_method(other, operator.lt)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ps_311_arrow_13_env\Lib\site-packages\pandas\core\series.py", line 5804, in _cmp_method
    res_values = ops.comparison_op(lvalues, rvalues, op)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ps_311_arrow_13_env\Lib\site-packages\pandas\core\ops\array_ops.py", line 332, in comparison_op
    res_values = op(lvalues, rvalues)
                 ^^^^^^^^^^^^^^^^^^^^
  File "C:\ps_311_arrow_13_env\Lib\site-packages\pandas\core\ops\common.py", line 76, in new_method
    return method(self, other)
           ^^^^^^^^^^^^^^^^^^^
  File "C:\ps_311_arrow_13_env\Lib\site-packages\pandas\core\arraylike.py", line 48, in __lt__
    return self._cmp_method(other, operator.lt)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\ps_311_arrow_13_env\Lib\site-packages\pandas\core\arrays\arrow\array.py", line 640, in _cmp_method
    raise NotImplementedError(
NotImplementedError: lt not implemented for <class 'pandas.core.arrays.arrow.array.ArrowExtensionArray'>
>>>
```

### Issue Description

The operation is implemented, the actual error should be an overflow from error about 9223372036854775808 not being in range of an int64, this is the error produced by pandas 2.0.3.

### Expected Behavior

I expect the error pandas 2.0.3 produced:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\pd_210\Lib\site-packages\pandas\core\ops\common.py", line 81, in new_method
    return method(self, other)
           ^^^^^^^^^^^^^^^^^^^
  File "C:\pd_210\Lib\site-packages\pandas\core\arraylike.py", line 48, in __lt__
    return self._cmp_method(other, operator.lt)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\pd_210\Lib\site-packages\pandas\core\series.py", line 6096, in _cmp_method
    res_values = ops.comparison_op(lvalues, rvalues, op)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\pd_210\Lib\site-packages\pandas\core\ops\array_ops.py", line 279, in comparison_op
    res_values = op(lvalues, rvalues)
                 ^^^^^^^^^^^^^^^^^^^^
  File "C:\pd_210\Lib\site-packages\pandas\core\ops\common.py", line 81, in new_method
    return method(self, other)
           ^^^^^^^^^^^^^^^^^^^
  File "C:\pd_210\Lib\site-packages\pandas\core\arraylike.py", line 48, in __lt__
    return self._cmp_method(other, operator.lt)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\pd_210\Lib\site-packages\pandas\core\arrays\arrow\array.py", line 467, in _cmp_method
    result = pc_func(self._data, other._data)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\pd_210\Lib\site-packages\pyarrow\compute.py", line 245, in wrapper
    return func.call(args, None, memory_pool)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pyarrow\_compute.pyx", line 367, in pyarrow._compute.Function.call
  File "pyarrow\error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow\error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Integer value 9223372036854775808 not in range: 0 to 9223372036854775807
>>>
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit              : ba1cccd19da778f0c3a7d6a885685da16a072870
python              : 3.11.2.final.0
python-bits         : 64
OS                  : Windows
OS-release          : 10
Version             : 10.0.22621
machine             : AMD64
processor           : Intel64 Family 6 Model 85 Stepping 7, GenuineIntel
byteorder           : little
LC_ALL              : None
LANG                : None
LOCALE              : English_United States.1252

pandas              : 2.1.0
numpy               : 1.25.2
pytz                : 2023.3
dateutil            : 2.8.2
setuptools          : 68.0.0
pip                 : 23.2.1
Cython              : None
pytest              : 7.0.1
hypothesis          : None
sphinx              : 4.2.0
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : None
html5lib            : None
pymysql             : None
psycopg2            : None
jinja2              : 3.1.2
IPython             : 8.14.0
pandas_datareader   : None
bs4                 : 4.12.2
bottleneck          : None
dataframe-api-compat: None
fastparquet         : None
fsspec              : None
gcsfs               : None
matplotlib          : None
numba               : None
numexpr             : None
odfpy               : None
openpyxl            : None
pandas_gbq          : None
pyarrow             : 13.0.0
pyreadstat          : None
pyxlsb              : None
s3fs                : None
scipy               : None
sqlalchemy          : None
tables              : None
tabulate            : 0.8.9
xarray              : None
xlrd                : None
zstandard           : None
tzdata              : 2023.3
qtpy                : None
pyqt5               : None
>>>

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [ ] closes #54944 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/arrays/arrow/array.py [hunk 1]
```diff
@@ -626,20 +626,22 @@ def __setstate__(self, state) -> None:
 
     def _cmp_method(self, other, op):
         pc_func = ARROW_CMP_FUNCS[op.__name__]
-        try:
+        if isinstance(other, (ArrowExtensionArray, np.ndarray, list, BaseMaskedArray)):
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:629]: Is it possible to split the cases when `ArrowNotImplementedError` vs `ArrowInvalid` is raised? I'm a little nervous of restricting the types of `other`?

Additionally maybe `self._box_pa` should be called first before going down the array like vs scalar path

Comment (phofl) [reply pandas/core/arrays/arrow/array.py:629]: this basically restores the logic we had before the last refactor that introduced the regression. I think calling box_pa on everything raised an exception, but can retry

**Comment by lithomas1** [comment]
Hm, this is kinda annoying so I think it should go into the release.

I'm not familiar enough to review this part of the code, though.

@mroeschke can you take another quick look

File: pandas/core/arrays/arrow/array.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/arrays/arrow/array.py:630]: Okay if this was the logic before I'm okay having this as a patch for 2.1.1, thought I think this should be more generic in the future.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @phofl

**Pull request merged**